### PR TITLE
Task/multiple roles groups

### PR DESCRIPTION
**Pull request by chriswk**
https://linear.app/unleash/issue/2-1128/change-the-api-to-support-adding-multiple-roles-to-a-usergroup-on-a
https://linear.app/unleash/issue/2-1125/be-able-to-fetch-all-roles-for-a-user-in-a-project
https://linear.app/unleash/issue/2-1127/adapt-the-ui-to-be-able-to-do-a-multi-select-on-role-permissions-for

- Allows assigning project roles to groups with root roles
- Implements new methods that support assigning, editing, removing and retrieving multiple project roles in project access, along with other auxiliary methods
- Adds new events for updating and removing assigned roles
- Adapts `useProjectApi` to new methods that use new endpoints that support multiple roles
- Adds the `multipleRoles` feature flag that controls the possibility of selecting multiple roles on the UI
- Adapts `ProjectAccessAssign` to support multiple role, using the new methods
- Adds a new `MultipleRoleSelect` component that allows you to select multiple roles based on the `RoleSelect` component
- Adapts the `RoleCell` component to support either a single role or multiple roles
- Updates the `access.spec.ts` Cypress e2e test to reflect our new logic
- Updates `access-service.e2e.test.ts` with tests covering the multiple roles logic and covering some corner cases
- Updates `project-service.e2e.test.ts` to adapt to the new logic, adding a test that covers adding access with `[roles], [groups], [users]`
- Misc refactors and boy scouting

image

### CODE REVIEW

File: src/lib/db/access-store.ts [hunk 1]
```diff
@@ -144,7 +144,6 @@ export class AccessStore implements IAccessStore {
                 .join(`${T.GROUP_ROLE} AS gr`, 'gu.group_id', 'gr.group_id')
                 .join(`${T.ROLE_PERMISSION} AS rp`, 'rp.role_id', 'gr.role_id')
                 .join(`${T.PERMISSIONS} AS p`, 'p.id', 'rp.permission_id')
-                .whereNull('g.root_role_id')
```
Comment (nunogois) [inline src/lib/db/access-store.ts:147]: Why are we deleting this line?

Comment (chriswk) [reply src/lib/db/access-store.ts:147]: Because otherwise, we won't be able to find project roles for groups with root roles.

File: frontend/cypress/integration/projects/access.spec.ts [hunk 2]
```diff
@@ -128,12 +142,31 @@ describe('project-access', () => {
...
+
+        cy.get(`[data-testid='${PA_ROLE_ID}']`).click();
+        cy.contains('full control over the project').click({
+            force: true,
+        });
+
+        cy.get(`[data-testid='${PA_ASSIGN_CREATE_ID}']`).click();
+        cy.wait('@editAccess');
+        cy.get("td span:contains('Owner')").should('have.length', 2);
+        cy.get("td span:contains('2 roles')").should('have.length', 1);
```
Comment (chriswk) [inline frontend/cypress/integration/projects/access.spec.ts:161]: :+1: , actual readable tests.

File: frontend/src/component/project/ProjectAccess/ProjectAccessAssign/ProjectAccessAssign.tsx [hunk 3]
```diff
@@ -430,11 +441,28 @@ export const ProjectAccessAssign = ({
...
+                                elseShow={() => (
+                                    <RoleSelect
+                                        data-testid={PA_ROLE_ID}
+                                        roles={roles}
+                                        value={selectedRoles[0]}
+                                        setValue={role =>
+                                            setRoles(role ? [role] : [])
+                                        }
+                                    />
+                                )}
```
Comment (chriswk) [inline frontend/src/component/project/ProjectAccess/ProjectAccessAssign/ProjectAccessAssign.tsx:465]: Nice way of using the flag to turn it on/off :)

File: frontend/src/hooks/api/actions/useProjectApi/useProjectApi.ts [hunk 4]
```diff
@@ -116,93 +117,59 @@ const useProjectApi = () => {
...
         const req = createRequest(path, { method: 'DELETE' });
 
-        try {
-            const res = await makeRequest(req.caller, req.id);
-
-            return res;
-        } catch (e) {
-            throw e;
-        }
+        return await makeRequest(req.caller, req.id);
```
Comment (chriswk) [inline frontend/src/hooks/api/actions/useProjectApi/useProjectApi.ts:142]: Good boyscouting here, why try-catch when we're just rethrowing :)
